README.md
transit-ruby
===================

Transit is a data format and a set of libraries for conveying
values between applications written in different languages. This
library provides support for marshalling Transit data to/from Ruby.

[Rationale](http://blog.cognitect.com/blog/2014/7/22/transit)<br>
[API docs](http://rubydoc.info/gems/transit-ruby)<br>
[Specification](https://github.com/cognitect/transit-format)

This implementation's major.minor version number corresponds to the
version of the Transit specification it supports.

_NOTE: Transit is a work in progress and may evolve based on feedback.
As a result, while Transit is a great option for transferring data
between applications, it should not yet be used for storing data
durably over time. This recommendation will change when the
specification is complete._

## Contributing

This library is open source, developed internally by Cognitect. We welcome discussions of potential problems and enhancement suggestions on the [transit-format mailing list](https://groups.google.com/forum/#!forum/transit-format). Issues can be filed using GitHub [issues](https://github.com/cognitect/transit-ruby/issues) for this project. Because transit is incorporated into products and client projects, we prefer to do development internally and are not accepting pull requests or patches.

## Releases and Dependency Information

See https://rubygems.org/gems/transit-ruby

## Install

```sh
gem install transit-ruby
```

## Basic Usage

```ruby
# io can be any Ruby IO

writer = Transit::Writer.new(:json, io) # or :json-verbose, :msgpack
writer.write(value)

reader = Transit::Reader.new(:json, io) # or :msgpack
reader.read

# or

reader.read {|val| do_something_with(val)}
```

For example:

```
irb(2.1.1): io = StringIO.new('', 'w+')
==========> #<StringIO:0x007faab2ec3970>
irb(2.1.1): writer = Transit::Writer.new(:json, io)
==========> #<Transit::Writer:0x007faab2e8c1c8 @marshaler=#<Transit::JsonMarshaler:0x007faab2e1a168..........(snip)..........
irb(2.1.1): writer.write("abc")
==========> nil
irb(2.1.1): writer.write(123456789012345678901234567890)
==========> nil
irb(2.1.1): io.string
==========> "[\"~#'\",\"abc\"]\n[\"~#'\",\"~n123456789012345678901234567890\"]\n"
irb(2.1.1): reader = Transit::Reader.new(:json, StringIO.new(io.string))
==========> #<Transit::Reader:0x007faab2db48e0 @reader=#<Transit::JsonUnmarshaler:0x007faab2dae030 @..........(snip)..........
irb(2.1.1): reader.read {|val| puts val}
abc
123456789012345678901234567890
```

## Custom Handlers

### Custom Write Handlers

Implement `tag`, `rep(obj)` and `string_rep(obj)` methods. For example:

```ruby
Point = Struct.new(:x,:y) do
  def to_a; [x,y] end
end

class PointWriteHandler
  def tag(_) "point" end
  def rep(o) o.to_a  end
  def string_rep(_) nil end
end
```

### Custom Read Handlers

Implement `from_rep(rep)` method. For example:

```ruby
class PointReadHandler
  def from_rep(rep)
    Point.new(*rep)
  end
end
```

### Example Usage

```ruby
io = StringIO.new('', 'w+')
writer = Transit::Writer.new(:json, io,
                             :handlers => {Point => PointWriteHandler.new})
writer.write(Point.new(37,42))

p io.string.chomp
#=> "[\"~#point\",[37,42]]"

reader = Transit::Reader.new(:json, StringIO.new(io.string),
                             :handlers  => {"point" => PointReadHandler.new})
p reader.read
#=> #<struct Point x=37, y=42>
```

See
[Transit::WriteHandlers](http://rubydoc.info/gems/transit-ruby/Transit/WriteHandlers)
for more info.

## Default Type Mapping

|Transit type|Write accepts|Read returns|Example(write)|Example(read)|
|------------|-------------|------------|--------------|-------------|
|null|nil|nil|nil|nil|
|string|String|String|"abc"|"abc"|
|boolean|true, false|true, false|false|false|
|integer|Fixnum, Bignum|Fixnum, Bignum|123|123|
|decimal|Float|Float|123.456|123.456|
|keyword|Symbol|Symbol|:abc|:abc|
|symbol|Transit::Symbol|Transit::Symbol|Transit::Symbol.new("foo")|`#<Transit::Symbol "foo">`|
|big decimal|BigDecimal|BigDecimal|BigDecimal.new("2**64")|`#<BigDecimal:7f9e6d33c558>`|
|big integer|Fixnum, Bignum|Fixnum, Bignum|2**128|340282366920938463463374607431768211456|
|time|DateTime, Date, Time|DateTime|DateTime.now|`#<DateTime: 2014-07-15T15:52:27+00:00 ((2456854j,57147s,23000000n),+0s,2299161j)>`|
|uri|Addressable::URI, URI|Addressable::URI|Addressable::URI.parse("http://example.com")|`#<Addressable::URI:0x3fc0e20390d4 URI:http://example.com>`|
|uuid|Transit::UUID|Transit::UUID|Transit::UUID.new|`#<Transit::UUID "defa1cce-f70b-4ddb-bb6e-b6ac817d8bc8">`|
|char|Transit::TaggedValue|String|Transit::TaggedValue.new("c", "a")|"a"|
|array|Array|Array|[1, 2, 3]|[1, 2, 3]|
|list|Transit::TaggedValue|Array|Transit::TaggedValue.new("list", [1, 2, 3])|[1, 2, 3]|
|set|Set|Set|Set.new([1, 2, 3])|`#<Set: {1, 2, 3}>`|
|map|Hash|Hash|`{a: 1, b: 2, c: 3}`|`{:a=>1, :b=>2, :c=>3}`|
|bytes|Transit::ByteArray|Transit::ByteArray|Transit::ByteArray.new("base64")|base64|
|link|Transit::Link|Transit::Link|Transit::Link.new(Addressable::URI.parse("http://example.org/search"), "search")|`#<Transit::Link:0x007f81c405b7f0 @values={"href"=>#<Addressable::URI:0x3fc0e202dfb8 URI:http://example.org/search>, "rel"=>"search", "name"=>nil, "render"=>nil, "prompt"=>nil}>`|

### Additional types (not required by the [transit-format](https://github.com/cognitect/transit-format) spec)

|Semantic type|Write accepts|Read returns|Example(write)|Example(read)|
|------------|-------------|------------|--------------|-------------|
|ratio|Rational|Rational|Rational(1, 3)|Rational(1, 3)|

## Supported Rubies

* MRI 1.9.3, 2.0.0, 2.1.0, 2.1.1, 2.1.2
* JRuby 1.7.13..16

## Copyright and License

Copyright © 2014 Cognitect

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
### 0.8.591 / 2015-05-03

* Bump lock_jar dependency to ~> 0.12.0 #17

### 0.8.588 / 2015-04-10

* Update to transit-java-0.8.287 for json int boundary fix in JRuby

### 0.8.586 / 2015-03-13

* Add handler caching for MRI
* Bump to transit-java-0.8.285 for handler caching in JRuby

### 0.8.572 / 2015-01-15

* Marshal int map keys as ints in msgpack

### 0.8.569 / 2014-12-03

* ByteArray#to_s forces default encoding for platform
  * fixes rare bug in which trying to print binary data nested within
    decoded binary data raises an encoding incompatibility error.

### 0.8.567 / 2014-09-21

* restore newline suppression when writing in json mode
* helpful error message when nested object has no handler

### 0.8.560 (java platform only) / 2014-09-12

* Bump dependency on transit-java to 0.8.269
  * fixes bug which turned an empty set into an array

### 0.8.552 (java platform only) / 2014-09-12

* Support JRuby!

### 0.8.539 / 2014-09-05

* Support special numbers (NaN, INF, -INF)

### 0.8.467 / 2014-07-22

* Initial release


lib/transit.rb
# Copyright 2014 Cognitect. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS-IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


# WriteHandlers convert instances of Ruby types to their corresponding Transit
# semantic types, and ReadHandlers read convert transit values back into instances
# of Ruby types. transit-ruby ships with default sets of WriteHandlers for each
# of the Ruby types that map naturally to transit types, and ReadHandlers for each
# transit type. For the common case, the
# built-in handlers will suffice, but you can add your own extension types and/or
# override the built-in handlers.
#
# For example, Ruby has Date, Time, and DateTime, each with their
# own semantics. Transit has an instance type, which does not
# differentiate between Date and Time, so transit-ruby writes Dates,
# Times, and DateTimes as transit instances, and reads transit
# instances as DateTimes. If your application cares that Dates are
# different from DateTimes, you could register custom write and read
# handlers, overriding the built-in DateHandler and adding a new DateReadHandler.
#
# ```ruby
# class DateWriteHandler
#   def tag(_) "D" end
#   def rep(o) o.to_s end
#   def string_rep(o) o.to_s end
# end
#
# class DateReadHandler
#   def from_rep(rep)
#     Date.parse(rep)
#   end
# end
#
# io = StringIO.new('','w+')
# writer = Transit::Writer.new(:json, io, :handlers => {Date => DateWriteHandler.new})
# writer.write(Date.new(2014,7,22))
# io.string
# # => "[\"~#'\",\"~D2014-07-22\"]\n"
#
# reader = Transit::Reader.new(:json, StringIO.new(io.string), :handlers => {"D" => DateReadHandler.new})
# reader.read
# # => #<Date: 2014-07-22 ((2456861j,0s,0n),+0s,2299161j)>
# ```
module Transit
  ESC = "~"
  SUB = "^"
  RES = "`"
  TAG = "~#"
  MAP_AS_ARRAY = "^ "
  TIME_FORMAT  = "%FT%H:%M:%S.%LZ"
  QUOTE = "'"

  MAX_INT = 2**63 - 1
  MIN_INT = -2**63

  JSON_MAX_INT = 2**53 - 1
  JSON_MIN_INT = -JSON_MAX_INT

  def jruby?
    defined?(RUBY_ENGINE) && RUBY_ENGINE == "jruby"
  end
  module_function :jruby?
end

require 'set'
require 'time'
require 'uri'
require 'base64'
require 'bigdecimal'
require 'securerandom'
require 'forwardable'
require 'addressable/uri'
require 'transit/date_time_util'
require 'transit/transit_types'
require 'transit/rolling_cache'
require 'transit/write_handlers'
require 'transit/read_handlers'
require 'transit/marshaler/base'
require 'transit/writer'
require 'transit/decoder'
require 'transit/reader'

if Transit::jruby?
  require 'lock_jar'
  LockJar.lock(File.join(File.dirname(__FILE__), "..", "Jarfile"))
  LockJar.load
  require 'transit.jar'
  require 'jruby'
  com.cognitect.transit.ruby.TransitService.new.basicLoad(JRuby.runtime)
  require 'transit/marshaler/jruby/json'
  require 'transit/marshaler/jruby/messagepack'
else
  require 'transit/marshaler/cruby/json'
  require 'transit/marshaler/cruby/messagepack'
  require 'transit/unmarshaler/cruby/json'
  require 'transit/unmarshaler/cruby/messagepack'
end
